contracts/Delay.sol

```diff
@@ -62,108 +77,182 @@ contract Delay is AragonApp, IForwarder {
 
     /**
-    * @notice Cancel script execution with ID `_delayedScriptId`
-    * @param _delayedScriptId The ID of the script execution to cancel
+    * @dev Reject script execution, IDisputable interface conformance
+    * @param _delayedScriptId The ID of the script execution to reject
     */
-    function cancelExecution(uint256 _delayedScriptId) external scriptExists(_delayedScriptId) auth(CANCEL_EXECUTION_ROLE) {
-        delete delayedScripts[_delayedScriptId];
+    function _onDisputableActionRejected(uint256 _delayedScriptId) internal {
+        DelayedScript storage delayedScript = delayedScripts[_delayedScriptId];
+
+        delayedScript.delayedScriptStatus = DelayedScriptStatus.Cancelled;
+
+        emit ExecutionCancelled(_delayedScriptId, delayedScript.actionId);
+    }
 
-        emit ExecutionCancelled(_delayedScriptId);
+    /**
+    * @dev Void script execution, IDisputable interface conformance
+    * @param _delayedScriptId The ID of the script execution to void
+    */
+    function _onDisputableActionVoided(uint256 _delayedScriptId) internal {
+        _onDisputableActionAllowed(_delayedScriptId);
     }
 
     /**
     * @notice Execute the script with ID `_delayedScriptId`
     * @param _delayedScriptId The ID of the script to execute
     */
-    function execute(uint256 _delayedScriptId) external {
-        require(canExecute(_delayedScriptId), ERROR_CAN_NOT_EXECUTE);
-
-        DelayedScript memory delayedScript = delayedScripts[_delayedScriptId];
-        delete delayedScripts[_delayedScriptId];
+    function execute(uint256 _delayedScriptId) external nonReentrant {
+        DelayedScript storage delayedScript = delayedScripts[_delayedScriptId];
+        require(_canExecute(delayedScript), ERROR_CAN_NOT_EXECUTE);
 
         runScript(delayedScript.evmCallScript, new bytes(0), new address[](0));
-        emit ExecutedScript(_delayedScriptId);
+
+        (,,,,bool closed,,) = _ensureAgreement().getAction(delayedScript.actionId);
+        if (!closed) {
+            _closeAgreementAction(delayedScript.actionId);
+        }
+
+        delayedScript.delayedScriptStatus = DelayedScriptStatus.Executed;
+
+        emit ExecutedScript(_delayedScriptId, delayedScript.actionId);
     }
 
     /**
-    * @notice Return whether a script with ID #`_scriptId` can be executed
-    * @param _scriptId The ID of the script to execute
+    * @notice Close agreement action with ID `_delayedScriptId`
+    * @param _delayedScriptId The ID of the script to close the agreement action of
     */
-    function canExecute(uint256 _scriptId) public view returns (bool) {
-        bool withinExecutionWindow = getTimestamp64() > delayedScripts[_scriptId].executionTime;
-        bool isUnpaused = !_isExecutionPaused(_scriptId);
+    function closeAgreementAction(uint256 _delayedScriptId) external {
+        DelayedScript storage delayedScript = delayedScripts[_delayedScriptId];
+        require(_canExecute(delayedScript), ERROR_CAN_NOT_EXECUTE);
```

Comment:
I think it makes sense to call `canClose` here as well :)

However, I'm not sure if I'd expose an external function for this, is there a case where you would like to close an action but not execute it?

Reply:
Done.

My interpretation of this message was that I should keep this function: keybase://chat/disputables#general/223 I also think if this is available through the Agreements app it isn't needed here. The issue would be that we still want to be able to close the action even if the evmScript isn't executable, ie reverts.